pkg/utils/apply/scheme.go

```diff
@@ -0,0 +1,64 @@
+package apply
+
+import (
+	admissionregistrationv1 "k8s.io/api/admissionregistration/v1"
+	admissionregistrationv1beta1 "k8s.io/api/admissionregistration/v1beta1"
+	appsv1 "k8s.io/api/apps/v1"
+	appsv1beta1 "k8s.io/api/apps/v1beta1"
+	appsv1beta2 "k8s.io/api/apps/v1beta2"
+	authenticationv1 "k8s.io/api/authentication/v1"
+	authenticationv1beta1 "k8s.io/api/authentication/v1beta1"
+	authorizationv1 "k8s.io/api/authorization/v1"
+	authorizationv1beta1 "k8s.io/api/authorization/v1beta1"
+	autoscalingv1 "k8s.io/api/autoscaling/v1"
+	autoscalingv2beta1 "k8s.io/api/autoscaling/v2beta1"
+	batchv1 "k8s.io/api/batch/v1"
+	batchv1beta1 "k8s.io/api/batch/v1beta1"
+	batchv2alpha1 "k8s.io/api/batch/v2alpha1"
+	certificatesv1beta1 "k8s.io/api/certificates/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	extensionsv1beta1 "k8s.io/api/extensions/v1beta1"
+	networkingv1 "k8s.io/api/networking/v1"
+	policyv1beta1 "k8s.io/api/policy/v1beta1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	rbacv1alpha1 "k8s.io/api/rbac/v1alpha1"
+	rbacv1beta1 "k8s.io/api/rbac/v1beta1"
+	schedulingv1alpha1 "k8s.io/api/scheduling/v1alpha1"
+	storagev1 "k8s.io/api/storage/v1"
+	storagev1beta1 "k8s.io/api/storage/v1beta1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// this scheme is only used to determine whether a resource is supported by StrategicMergePatch.
+var scheme = runtime.NewScheme()
+var builder = runtime.SchemeBuilder{
+	admissionregistrationv1.AddToScheme,
+	admissionregistrationv1beta1.AddToScheme,
+	appsv1.AddToScheme,
+	appsv1beta1.AddToScheme,
+	appsv1beta2.AddToScheme,
+	authenticationv1.AddToScheme,
+	authenticationv1beta1.AddToScheme,
+	authorizationv1.AddToScheme,
+	authorizationv1beta1.AddToScheme,
+	autoscalingv1.AddToScheme,
+	autoscalingv2beta1.AddToScheme,
+	batchv1.AddToScheme,
+	batchv1beta1.AddToScheme,
+	batchv2alpha1.AddToScheme,
+	certificatesv1beta1.AddToScheme,
+	corev1.AddToScheme,
+	extensionsv1beta1.AddToScheme,
+	networkingv1.AddToScheme,
+	policyv1beta1.AddToScheme,
+	rbacv1.AddToScheme,
+	rbacv1beta1.AddToScheme,
+	rbacv1alpha1.AddToScheme,
+	schedulingv1alpha1.AddToScheme,
+	storagev1beta1.AddToScheme,
+	storagev1.AddToScheme,
+}
+
+func init() {
+	_ = builder.AddToScheme(scheme)
+}
```

Comment:
You don't need to manually add the built-in scheme here. client-go has already done it for you:

Check `k8s.io/client-go/kubernetes/scheme`

Reply:
Yes, I tried to use client-go scheme directly. But later I found not all types registered in client-go scheme are supported by strategic-merge as kubectl shown. so I just pick up the supported types from client-go scheme.

And another reason why not use client-go scheme is client-go/kubernetes/scheme.Scheme is a pkg level pointer variable.  It's possible to inject other types into it accidently, e.g., the test in /test/integration injects vela scheme into client-go scheme.Scheme indirectly.

Reply:
Does kubectl client code use their own?

Reply:
Yes, it uses [this scheme](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubectl/pkg/scheme/install.go) to determine strategicmerge or jsonmerge.

Reply:
Can we use the same package with the kubectl's?

Reply:
I checked again. Kubectl also use the scheme of client-go. I make a copy of client-go scheme to use now, instead of using the scheme of client-go directly.